src/backend/distributed/citus--8.3-1--8.4-1.sql

```diff
@@ -1,6 +1,105 @@
 /* citus--8.3-1--8.4-1 */
 
 /* bump version to 8.4-1 */
+CREATE SCHEMA IF NOT EXISTS citus_internal;
+
+-- move citus internal functions to citus_internal to make space in the citus schema for
+-- our public interface
+ALTER FUNCTION citus.find_groupid_for_node SET SCHEMA citus_internal;
+ALTER FUNCTION citus.pg_dist_node_trigger_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.pg_dist_shard_placement_trigger_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.refresh_isolation_tester_prepared_statement SET SCHEMA citus_internal;
+ALTER FUNCTION citus.replace_isolation_tester_func SET SCHEMA citus_internal;
+ALTER FUNCTION citus.restore_isolation_tester_func SET SCHEMA citus_internal;
+
+CREATE OR REPLACE FUNCTION citus_internal.pg_dist_shard_placement_trigger_func()
+RETURNS TRIGGER AS $$
+  BEGIN
+    IF (TG_OP = 'DELETE') THEN
+      DELETE FROM pg_dist_placement WHERE placementid = OLD.placementid;
+      RETURN OLD;
+    ELSIF (TG_OP = 'UPDATE') THEN
+      UPDATE pg_dist_placement
+        SET shardid = NEW.shardid, shardstate = NEW.shardstate,
+            shardlength = NEW.shardlength, placementid = NEW.placementid,
+            groupid = citus_internal.find_groupid_for_node(NEW.nodename, NEW.nodeport)
+        WHERE placementid = OLD.placementid;
+      RETURN NEW;
+    ELSIF (TG_OP = 'INSERT') THEN
+      INSERT INTO pg_dist_placement
+        (placementid, shardid, shardstate, shardlength, groupid)
+      VALUES (NEW.placementid, NEW.shardid, NEW.shardstate, NEW.shardlength,
+        citus_internal.find_groupid_for_node(NEW.nodename, NEW.nodeport));
+      RETURN NEW;
+    END IF;
+  END;
+$$ LANGUAGE plpgsql;
+
+CREATE TABLE citus.pg_dist_object (
+    classid oid NOT NULL,
+    objid oid NOT NULL,
+    objsubid integer NOT NULL,
+
+    -- fields used for upgrades
+    type text DEFAULT NULL,
+    object_names text[] DEFAULT NULL,
+    object_args text[] DEFAULT NULL
+);
+
+CREATE INDEX pg_dist_object_classid_objid_objsubid_index ON
+    citus.pg_dist_object USING btree(classid, objid, objsubid);
+
+GRANT USAGE ON SCHEMA citus TO public;
+GRANT SELECT ON citus.pg_dist_object TO public;
+GRANT DELETE ON citus.pg_dist_object TO public;
```

Comment:
This seems too permissive, public is all users. Maybe remove this line.

Reply:
Probably worth pointing out that these permissions only apply to the SQL interface, not when using C functions.

Reply:
Maybe also remove the SELECT line, since this came up for other tables in the security audit.

Reply:
I think we can drop the select statement, but the `DELETE` has been added deliberately for the `citus_drop_trigger`. This gets run as any user during a `DROP CASCADE`, this failed if the user had no `DELETE` rights on pg_dist_object.

The only way to get the grants gone is to rewrite that back to `C`.

Reply:
Note; if we write this logic in C it still needs to be callable by public. Which is probably the same for functions like:
 - `master_remove_distributed_table_metadata_from_workers`
 - `master_drop_all_shards`
 - `master_remove_partition_metadata`
 - `master_drop_sequences`

Which seem pretty destructive. Guess they have some ownership checks on them, which we would need as well for the object address'

Reply:
Those functions check ownership in CheckTableSchemaNameForDrop if the table still exists (called directly), or they will error out on the workers if the table no longer exists on the coordinator (usually, because they are called from the drop trigger). 

We cannot really allow arbitrary users to delete from metadata tables. One solution would be to move the DELETE to a security definer function (which is executed as superuser), which deletes the objects from pg_dist_object *only if they no longer exist*. That way, it just becomes a clean-up function that anyone can call, but will never do anything destructive.

Reply:
That sounds like a good option, unfortunately I can't find a function that can check based on an ObjectAddress if it exists or not. We would need to implement that for every postgres ObjectClass.

Alternatively there is a function that can do an ownership check, but that takes more arguments which are dependant on what kind of object you are checking, causing it non-trivial to call. Also, in the drop trigger I guess the object is already deleted so an ownership check is impossible.

I will work out a function that can check if an object still exists, and remove it from `pg_dist_object` if it is non-existing anymore. Before creating the existence check I will have an extra round throughout postgres to see if there is something like it around.

Reply:
https://github.com/citusdata/citus/pull/2882/commits/a05e24c437933657c450a4618bc9d6cf2b2de231